docs/JAX_machine_translation.md

```diff
@@ -0,0 +1,502 @@
+```
+
+```{code-cell} ipython3
+tokenizer = tiktoken.get_encoding("cl100k_base")
+```
+
+We strip out puctuation to keep things simple and in line with the original tutorial - the `[` `]` are kept in so that our `[start]` and `[end]` formatting is preserved.
+
+```{code-cell} ipython3
+strip_chars = string.punctuation + "¿"
+strip_chars = strip_chars.replace("[", "")
+strip_chars = strip_chars.replace("]", "")
+
+vocab_size = tokenizer.n_vocab
+sequence_length = 20
+```
+
+```{code-cell} ipython3
+def custom_standardization(input_string):
+    lowercase = input_string.lower()
+    return re.sub(f"[{re.escape(strip_chars)}]", "", lowercase)
+```
+
+```{code-cell} ipython3
+def tokenize_and_pad(text, tokenizer, max_length):
+    tokens = tokenizer.encode(text)[:max_length]
+    padded = tokens + [0] * (max_length - len(tokens)) if len(tokens) < max_length else tokens ##assumes list-like - (https://github.com/openai/tiktoken/blob/main/tiktoken/core.py#L81 current tiktoken out)
+    return padded
+```
+
+```{code-cell} ipython3
+def format_dataset(eng, spa, tokenizer, sequence_length):
+    eng = custom_standardization(eng)
+    spa = custom_standardization(spa)
+    eng = tokenize_and_pad(eng, tokenizer, sequence_length)
+    spa = tokenize_and_pad(spa, tokenizer, sequence_length)
+    return {
+            "encoder_inputs": eng,
+            "decoder_inputs": spa[:-1],
+            "target_output": spa[1:],
+            }
+```
+
+```{code-cell} ipython3
+train_data = [format_dataset(eng, spa, tokenizer, sequence_length) for eng, spa in train_pairs]
+val_data = [format_dataset(eng, spa, tokenizer, sequence_length) for eng, spa in val_pairs]
+test_data = [format_dataset(eng, spa, tokenizer, sequence_length) for eng, spa in test_pairs]
+```
+
+At this point we've extracted the data, applied formatting, and tokenized the phrases with padding. The data is kept in train/validate/test sets that each have dictionary entries, which look like the following:
+
+```{code-cell} ipython3
+## data selection example
+print(train_data[135])
+```
+
+## Define Transformer components: Encoder, Decoder, Positional Embed
+
+In many ways this is very similar to the original source, with `ops` changing to `jnp` and `keras` or `layers` becoming `nnx`. Certain module-specific arguments come and go, like the rngs attached to most things in the updated version, and decode=False in the MHA call.
+
+```{code-cell} ipython3
+class TransformerEncoder(nnx.Module):
+    def __init__(self, embed_dim: int, dense_dim: int, num_heads: int, rngs: nnx.Rngs, **kwargs):
+        self.embed_dim = embed_dim
+        self.dense_dim = dense_dim
+        self.num_heads = num_heads
+
+        self.attention = nnx.MultiHeadAttention(num_heads=num_heads,
+                                          in_features=embed_dim,
+                                          decode=False,
+                                          rngs=rngs) 
+        self.dense_proj = nnx.Sequential(
+            *[
```

Comment:
Technically, you do not need `*` and `[` `]`, just provide layers with comma between them